### PR TITLE
Implement AppThemeColor, AppThemeObject, and AppThemeResource

### PR DESCRIPTION
**Pull request by jfversluis**
Still todo:
- [x] Determine if we even want this (I think we do)
- [x] Find the right place/namespace for it
- [x] Add some samples for it
- [x] Add code documentation
- [x] Add conceptual docs: https://github.com/MicrosoftDocs/CommunityToolkit/pull/297

This adds the AppThemeColor and AppThemeResource objects that make it easier to work with colors and other resources in light/dark theme scenarios. This enables you to define colors (and resources) in resource dictionaries, making your XAML less verbose and more reusable.

Original PR on .NET MAUI by Dan Siegel: https://github.com/dotnet/maui/pull/8839

### CODE REVIEW

File: src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs
Comment (jfversluis) [inline src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs:7]: We probably want to rename this because now the extension is `AppTheme` which collides with the `AppTheme` in .NET MAUI and Visual Studio suggests the wrong namespace to be imported. Unless you use the full name `AppThemeExtension` but not sure if that's desirable.

File: src/CommunityToolkit.Maui/AppThemeColor.shared.cs
Comment (jfversluis) [inline src/CommunityToolkit.Maui/AppThemeColor.shared.cs:1]: What's going to be the right namespace for this?

File: src/CommunityToolkit.Maui/AppThemeResource.shared.cs
Comment (jfversluis) [inline src/CommunityToolkit.Maui/AppThemeResource.shared.cs:1]: What's going to be the right namespace for this?

File: src/CommunityToolkit.Maui/AppThemeColor.shared.cs
Comment (jfversluis) [inline src/CommunityToolkit.Maui/AppThemeColor.shared.cs:6]: With some effort we could make this an inheritance from `AppThemeResource`? Not sure why that hasn't been done.

**Review by brminnick — CHANGES_REQUESTED** [review]
Hey Gerald! I've only done a partial review, but at first glance it looks like we can simplify `AppThemeColor.cs` a bit. What do you think?

File: samples/CommunityToolkit.Maui.Sample/Pages/Essentials/AppThemePage.xaml
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/Pages/Essentials/AppThemePage.xaml:6]: Let's use toolkit to not confuse people it's a MAUI namepaspace

File: src/CommunityToolkit.Maui.UnitTests/AppThemeTests.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.UnitTests/AppThemeTests.cs:8]: Let's move it to all other mocks

**Comment by jfversluis** [comment]
@brminnick yep thanks! That was my outstanding comment on refactoring it to use the same code. Applied that now. There is one base `AppThemeObject<T>` and 2 convenience objects basically that inherit from it, but don't add anything except for making it easier to work with in both code and XAML: `AppThemeResource` which just uses an `object` and `AppThemeColor` which does the same for colors. What do you think?

I've left `AppThemeObject<T>` public on purpose, maybe people want to extend it for their own things?

File: src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs:85]: Can we remove this ToDo?

File: src/CommunityToolkit.Maui/AppThemeColor.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/AppThemeColor.shared.cs:6]: Can we make this `sealed`, can't see any value allowing users to inherit from it

```suggestion
public sealed class AppThemeColor : AppThemeObject<Color>
```

File: src/CommunityToolkit.Maui/AppThemeResource.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/AppThemeResource.shared.cs:6]: Can we make this `sealed` for the same reason as the `AppThemeColor`?

```suggestion
public sealed class AppThemeResource : AppThemeObject<object>
```

**Review by pictos — COMMENTED** [review]
This looks good in general, I've feel comments around the `AppThemeObject` class

File: src/CommunityToolkit.Maui/AppThemeObjectOfT.shared.cs
Comment (jfversluis) [reply src/CommunityToolkit.Maui/AppThemeObjectOfT.shared.cs:6]: Yeah I thought about this later as well, I think this is the right move! Thanks!

File: src/CommunityToolkit.Maui/AppThemeObjectOfT.shared.cs
Comment (jfversluis) [reply src/CommunityToolkit.Maui/AppThemeObjectOfT.shared.cs]: You would think so! However this is needed. If we don't do this it won't work correctly. For instance: when you only specify a color for Default and Dark theme and you set your device to Light theme, instead of Default, it will use Light which is then null.

**Comment by jfversluis** [comment]
OK I think I have all the feedback processed. Then the only question remains if we want to rename `AppThemeExtension` to anything.

If you start typing `<Label TextColor="{AppTheme ...` it will suggest the [`AppTheme`](https://github.com/dotnet/maui/blob/main/src/Essentials/src/AppInfo/AppTheme.shared.cs#L6) type in .NET MAUI. Is that a problem? I think ideally we'd pick a different name to avoid collisions. On the other hand, people will most likely add the xmlns first and then type `toolkit:AppTheme...` and it will suggest the right one?

File: samples/CommunityToolkit.Maui.Sample/Pages/Essentials/AppThemePage.xaml
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/Pages/Essentials/AppThemePage.xaml:23]: Can we add a theme toggle button?

File: src/CommunityToolkit.Maui/CommunityToolkit.Maui.csproj
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/CommunityToolkit.Maui.csproj:80]: All changes in this file can be reverted

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs:32]: Do we need a check for null? Is null a valid value? What if null is a value value for T and I want to set it after the previous value?

File: src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs:20]: Use ArgumentException.ThrowIfNull

File: src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs:27]: Does sourceProvider a valid parameter?

File: src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs:30]: Can we use generic GetService or GetRequiredService?

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs:32]: Already answered here
https://github.com/CommunityToolkit/Maui/pull/1264#discussion_r1241127495

File: src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui/Extensions/AppThemeExtension.shared.cs]: Yup, it's a valid parameter for `XamlParseException`

**Comment by brminnick** [comment]
> Then the only question remains if we want to rename AppThemeExtension to anything.

We could change it to `BindingBaseExtension` to align with our existing `IMarkupExtension` classes; their class name references the IMarkupExtension Type (eg `ValueConverterExtension ` implements `IMarkupExtension<ICommunityToolkitValueConverter> `)
- `public class MultiValueConverterExtension : IMarkupExtension<ICommunityToolkitMultiValueConverter>`
- `public abstract class ValueConverterExtension : BindableObject, IMarkupExtension<ICommunityToolkitValueConverter>`


```cs
public sealed class AppThemeExtension : IMarkupExtension<BindingBase>
```
changes to 
```cs
public sealed class BindingBaseExtension : IMarkupExtension<BindingBase>
```

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeColor.shared.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeColor.shared.cs:1]: But this is a `xaml` feature, so because of that it shouldn't go to `Core`

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs:32]: This may be an issue if null is a valid value. We cannot set it after some "not null" value

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeColor.shared.cs
Comment (jfversluis) [reply src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeColor.shared.cs:1]: I don't think .NET MAUI Core has all the references we need to make this happen. I put it in core first and it couldn't find all the references to types it needed.

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs:1]: But this code has a reference to the Binding system, which is MVVM oriented. The core assembly should be agnostic

**Comment by IndianaGary** [comment]
Re Brandon's chosen rename for `AppThemeExtension`: It occurs to me that `BindingBaseExtension `isn't really appropriate as it doesn't identify it's purpose and is too generic. I suggest changing it to `ThemedColorExtension`, as it concisely specifies its purpose and doesn't clash with other names.

**Comment by jfversluis** [comment]
Thanks for the input @IndianaGary however, this isn't just about colors. It's also resources in general.

So `{ThemedColor MyResource}` might make people think that it's only for colors, which it's not.

In our core maintainers chat I opted for just Theme? `TextColor={Theme MyColorThing}` and `ImageSource={Theme MyImage}`.

However, that maybe makes the names `AppThemeColor` and `AppThemeResource` a bit weird maybe. Should they be `ThemeColor` and `ThemeResource` then? But that will make them different from `AppThemeBinding` (in .NET MAUI) that they're based on.

Or maybe at this point I'm overthinking it 😄

**Comment by IndianaGary** [comment]
That's possible... ii was being guided by the specificity of AppThemeColor and AppThemeResource being distinct markup extensions and the issue of name collisions. ThemedColor IS targeted specifically at color. Similarly, Themed Resource would reference a resource. Trying to come up with a combined extension violates the Single Responsibility principle hence my preference for a targeted name.

**Comment by jfversluis** [comment]
>  Similarly, Themed Resource would reference a resource.

A color is a resource though, even in this code, it literally just inherits from the resource. But this is more about it being clear to the consumer and not the technicalities I guess 😄

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeObjectOfT.shared.cs:6]: Shall we rename this to `ThemeObject<T>`?

After renaming `ThemeResourceExtension`, it's a bit confusing to be using `<mct:AppThemeColor ...`, `<mct:AppThemeResource ...`, and `{mct:ThemeResource ...` in XAML. 

I assume devs would prefer consistency of each being named `Theme*`

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeColor.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeColor.shared.cs:6]: Shall we rename this to `ThemeColor`?

After renaming `ThemeResourceExtension`, it's a bit confusing to be using `<mct:AppThemeColor ...`, `<mct:AppThemeResource ...`, and `{mct:ThemeResource ...` in XAML. 

I assume devs would prefer consistency of each being named `Theme*`

File: src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeResource.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Essentials/AppTheme/AppThemeResource.shared.cs:6]: Shall we rename this to `ThemeResource`?

After renaming `ThemeResourceExtension`, it's a bit confusing to be using `<mct:AppThemeColor ...`, `<mct:AppThemeResource ...`, and `{mct:ThemeResource ...` in XAML. 

I assume devs would prefer consistency of each being named `Theme*`

**Review by brminnick — COMMENTED** [review]
Hey Gang! I think we should rename everything to `Theme...` now that we renamed `ThemeResouceExtension`.

Renaming will give us this consistency in XAML:
- Updated: `<mct:ThemeResource ... />
  - Current  `<mct:AppThemeResource ... />
- Updated: `<mct:ThemeColor ... />
  - Current: `<mct:AppThemeColor ... />
- Current: `{mct:ThemeResource ...}`

**Comment by jfversluis** [comment]
@brminnick So yeah, funny story, now we're back to square 1.

The extension is called: ThemeResource and now we have an object ThemeResource... 😅

**Comment by brminnick** [comment]
Oh no!! 🙈🙃😭

Naming is so hard...

image

**Comment by IndianaGary** [comment]
😊

**Review by brminnick — APPROVED** [review]
Thanks Gerald!! I think we finally got the naming:

```cs
public sealed class AppThemeColor : AppThemeObject<Color>;

public sealed class AppThemeObject : AppThemeObject<object>;

public abstract class AppThemeObject<T>;

public sealed class AppThemeResourceExtension : IMarkupExtension<BindingBase>
```

I approve, and I'll defer to you to merge it to give you time to review the updates.

:point_down: Here's how these names look in the Sample app

Screenshot 2023-07-23 at 1 16 38 PM

**Pull request merged**
